presto-main/src/main/java/com/facebook/presto/sql/planner/TimestampRuleChange.java

```diff
@@ -0,0 +1,144 @@
+import com.facebook.presto.spi.plan.FilterNode;
+import com.facebook.presto.spi.relation.CallExpression;
+import com.facebook.presto.spi.relation.ConstantExpression;
+import com.facebook.presto.spi.relation.RowExpression;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.relational.FunctionResolution;
+
+import java.time.LocalDate;
+import java.time.format.DateTimeParseException;
+import java.util.Optional;
+
+import static com.facebook.presto.common.function.OperatorType.GREATER_THAN_OR_EQUAL;
+import static com.facebook.presto.common.type.TimestampType.TIMESTAMP;
+import static com.facebook.presto.sql.planner.plan.Patterns.filter;
+import static com.facebook.presto.sql.relational.Expressions.comparisonExpression;
+import static com.facebook.presto.sql.relational.Expressions.constant;
+
+/**
+ * Minimal iterative rule that rewrites:
+ *   CAST(ts_col AS date) = DATE 'yyyy-mm-dd'
+ * into
+ *   ts_col >= TIMESTAMP 'yyyy-mm-dd 00:00:00.000'
+ * (No upper bound to avoid referencing "AND".)
+ */
+public class TimestampRuleChange
+        implements Rule<FilterNode>
+{
+    private static final Pattern<FilterNode> PATTERN = filter();
+
+    private final FunctionAndTypeManager functionAndTypeManager;
+    private final StandardFunctionResolution functionResolution;
+
+    public TimestampRuleChange(FunctionAndTypeManager functionAndTypeManager)
+    {
+        this.functionAndTypeManager = functionAndTypeManager;
+        this.functionResolution = new FunctionResolution(functionAndTypeManager.getFunctionAndTypeResolver());
+    }
+
+    @Override
+    public Pattern<FilterNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(FilterNode filter, Captures captures, Context context)
+    {
+        RowExpression originalPredicate = filter.getPredicate();
+        RowExpression rewritten = rewritePredicate(originalPredicate);
+        if (rewritten.equals(originalPredicate)) {
+            return Result.empty();
+        }
+
+        FilterNode newFilter = new FilterNode(
+                filter.getSourceLocation(),
+                filter.getId(),
+                filter.getSource(),
+                rewritten);
+        return Result.ofPlanNode(newFilter);
+    }
+
+    private RowExpression rewritePredicate(RowExpression expression)
+    {
+        // If not a function call, do nothing
+        if (!(expression instanceof CallExpression)) {
+            return expression;
+        }
+        CallExpression call = (CallExpression) expression;
+
+        // If the displayName is '=' and there are exactly two arguments, check for date(...) usage
+        if ("=".equalsIgnoreCase(call.getDisplayName()) && call.getArguments().size() == 2) {
+            RowExpression left = call.getArguments().get(0);
+            RowExpression right = call.getArguments().get(1);
+
+            Optional<RowExpression> maybeCol = extractCastToDate(left);
```

Comment:
Since we're looking for the `date` function, we might need to change the approach to extract the argument from the `date` function instead

Reply:
Also, the arguments could potentially be swapped, to be more robust, we should try to handle the case where maybeCol is executed against the `right` in case the `left` is not found (and similarly for the date constant)